prototype2/pubsub/src/lib.rs

```diff
@@ -0,0 +1,189 @@
+/// access to.
+pub struct SubscriberToken<T> {
+  phantom_data: PhantomData<T>,
+  reader_idx: u32,
+}
+
+struct PubSubRegistry {
+  readers: HashMap<TypeId, u32>,
+}
+
+
+impl PubSubRegistry {
+  fn new() -> PubSubRegistry {
+    PubSubRegistry { readers: HashMap::new() }
+  }
+}
+
+impl PubSubStore for specs::World {
+  /// Attempt to retrieve the general publisher for this resource
+  fn read_publisher<'a, T: Clone + Any + Send + Sync>(&'a mut self,
+                                                      token: &PublisherToken<T>)
+                                                      -> Publisher<'a, T> {
+
+    Publisher { senders: acquire_raw_containers::<T>(self) }
+  }
+
+  /// Attempt to retrieve the subscriber corresponding to this token
+  fn read_subscriber<'a, T: Clone + Any + Send + Sync>(&'a mut self,
+                                                       token: &SubscriberToken<T>)
+                                                       -> Subscriber<'a, T> {
+
+    Subscriber {
+      receivers: acquire_raw_containers::<T>(self),
+      idx: token.reader_idx,
+    }
+  }
+
+  /// Provide a publisher token for the user. We don't actually record this,
+  /// its merely to make
+  /// sure they follow procedure to get access to the resource.
+  fn register_publisher<T: Clone + Any + Send + Sync>(&mut self) -> PublisherToken<T> {
+    PublisherToken { phantom_data: PhantomData }
+  }
+
+  /// Allocate the subscriber a Receiver and provide them a token containing
+  /// that receiver index.
+  fn register_subscriber<T: Clone + Any + Send + Sync>(&mut self) -> SubscriberToken<T> {
+    let idx = {
+      let mut registry = acquire_registry(self);
+
+      let type_id = TypeId::of::<T>();
+      let mut idx = registry.readers.entry(type_id).or_insert(0);
+      *idx += 1;
+      idx.clone() - 1
+    };
+
+    acquire_raw_containers::<T>(self).push(NewContainer(Vec::new()));
+
+    SubscriberToken {
+      phantom_data: PhantomData,
+      reader_idx: idx,
+    }
+  }
+}
+
+// Wrap builtins so they cannot be accessed outside this module
+// This prevents accidental access by someone who happens to be storing Vecs
+struct NewContainer<T: Clone + Any + Send + Sync>(pub Vec<T>);
+
+impl<T: Clone + Any + Send + Sync> NewContainer<T> {
+  pub fn get_mut_vec(&mut self) -> &mut Vec<T> {
+    let &mut NewContainer(ref mut stuff) = self;
+    stuff
+  }
+}
+
+pub struct Publisher<'a, T: Any + Clone + Sync + Send> {
+  senders: RwLockWriteGuard<'a, Vec<NewContainer<T>>>,
+}
+
+impl<'a, T: Any + Clone + Sync + Send> Publisher<'a, T> {
+  pub fn push(&mut self, item: T) {
+    self.senders.iter_mut().foreach(|mut s| s.get_mut_vec().push(item.clone()))
+  }
+}
+
+pub struct Subscriber<'a, T: Any + Clone + Sync + Send> {
+  receivers: RwLockWriteGuard<'a, Vec<NewContainer<T>>>,
+  idx: u32,
+}
+
+impl<'a, T: Any + Clone + Sync + Send> Subscriber<'a, T> {
+  pub fn pull(&mut self) -> &mut Vec<T> {
+    self.receivers[self.idx as usize].get_mut_vec()
```

Comment:
Thats actually no good: this is not a mere facade for a vec: I've narrowed the API down to just `push` (and probably `extend` soon since thats a common pattern)